### PR TITLE
Add initial support for RPM packaging

### LINKED ISSUE #7: Package rpm (Fedora, CentOS, RedHad)
Sunshine does not currently have a method to package rpm.

If you have experience with packaging rpm and would like to contribute a PR would be very appreciated!

### PR DESCRIPTION
**Pull request by thatsysadmin**
## Description

This PR should add initial support for building an RPM package for the Fedora Linux distribution.

The Fedora containers should now build RPM packages as well.

### Issues Fixed or Closed

- Fixes #7 

## Type of Change

- [x] New feature (non-breaking change which adds functionality)

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have added or updated the documentation blocks for new or existing components

### CODE REVIEW

**Review by ReenigneArcher — CHANGES_REQUESTED** [review]
I've made a PR against your branch with the required changes for the workflow. Let me know your thoughts.

File: gen-rpm.in
Comment (thatsysadmin) [reply gen-rpm.in]: I forgot about that when I copy-pasted the script and changed it to use the cmake flags. Thanks for catching that.

File: scripts/build-private.sh
Comment (thatsysadmin) [reply scripts/build-private.sh:40]: I could change that.

Comment (thatsysadmin) [reply scripts/build-private.sh]: Just changed it now.

**Comment by thatsysadmin** [comment]
Ok. I pulled in your PR for my branch and modified the dockerfiles for the Ubuntu containers. Thanks for catching my mistakes.

**Review by ReenigneArcher — CHANGES_REQUESTED** [review]
Thanks for making those changes! The Debian Dockerfiles will need `-deb` added as well.

It also appears there are conflicts that need to be resolved before I can run tests or merge.

**Comment by thatsysadmin** [comment]
Ok. I had to rebase my master branch with the nightly. I was working completely with the master, so I didn't catch the Debian dockerfile as well that had to be changed that was in the nightly branch. Sorry about that.

**Comment by ReenigneArcher** [comment]
I think I missed this in the `build-sunshine.sh` thus causing the build to fail in the workflow.

https://github.com/thatsysadmin/Sunshine/blob/7cd40c4324a04398864b51568ee04cbf2208f509/scripts/build-sunshine.sh#L37

to

```bash
while getopts ":dpeuhc:s:n:" arg; do
```

**Comment by thatsysadmin** [comment]
@ReenigneArcher 

Do we need to build for Fedora 33? [Fedora 33 is EOL.](https://docs.fedoraproject.org/en-US/releases/eol/)

**Comment by ReenigneArcher** [comment]
> @ReenigneArcher
> 
> Do we need to build for Fedora 33? [Fedora 33 is EOL.](https://docs.fedoraproject.org/en-US/releases/eol/)

No, I was planning to remove it from the CI. You can take care of that if you like. Just remove the 3 lines from the CI.yml file in the matrix setup.

Still have the same issue on Fedora 35 though. Not quite sure what's wrong in the sunshine_build.sh file.

**Comment by thatsysadmin** [comment]
I swear. It feels like something has to always break overnight.

I ran the CI action on my own repository just now, and now it won't pass the Changelog check. It passed it last night. I'm not sure what I did.

**Comment by thatsysadmin** [comment]
@ReenigneArcher Sorry if this is a confusing question.

When the CI runs runs the Changelog verification, should there be anything in the `get_latest_release` [outputs? ](https://github.com/thatsysadmin/Sunshine/runs/5963255815?check_suite_focus=true) (output lines 54-59) The runs yesterday don't have anything in the `get_latest_release` [outputs.](https://github.com/thatsysadmin/Sunshine/runs/5958077460?check_suite_focus=true) This is sort of uncharted waters for me, so I might need some help in fixing this.

**Comment by ReenigneArcher** [comment]
@thatsysadmin no, that's sort of conditional. It's when a PR or push is made against `master`.

The values are used to set the next release version.

**Comment by ReenigneArcher** [comment]
I don't believe anything else in the CI needs to change at this point. Just in the build-sunshine.sh... it's something to do with the new argument that was added. The logic on the argument parsing didn't quite make sense to me to be honest, so I might have messed it up.

**Comment by thatsysadmin** [comment]
@ReenigneArcher 

Ok. I was triggering the CI to test the change to line 37.

**Comment by ReenigneArcher** [comment]
> @ReenigneArcher
> 
> Ok. I was attempting to trigger the CI on my fork to test the change to line 37 in the `build-sunshine.sh` script you mentioned.

Probably going to fail because you used `master` branch in your fork.

**Comment by ReenigneArcher** [comment]
I believe I found the problem.

Indentation needs to be increased here: https://github.com/SunshineStream/Sunshine/blob/505939c19f7229a721c6881d6211fd4607bc032f/scripts/build-sunshine.sh#L58-L62

**Comment by thatsysadmin** [comment]
> > @ReenigneArcher
> > Ok. I was attempting to trigger the CI on my fork to test the change to line 37 in the `build-sunshine.sh` script you mentioned.
> 
> Probably going to fail because you used `master` branch in your fork.

Yeah, I didn't realize that I needed to fork the nightly branch when I started this endeavor. I knew it would bite me.

**Comment by ReenigneArcher** [comment]
@thatsysadmin I need to update the documentation to make that more clear for contributors.

Also, sorry this is being stupid. Changes to CI are a pain to test.

Can you run the `build_sunshine.sh` locally?

**Comment by thatsysadmin** [comment]
@ReenigneArcher Got an idea.

Instead of using: (lines 55 - 62)
```
if `list_include_item "deb rpm" $OPTARG`
then
	SUNSHINE_PACKAGE_EXTENSION=$OPTARG
else
	echo "Package extension not supported: $OPTARG"
	echo "Falling back to default package extension: $SUNSHINE_PACKAGE_EXTENSION"
fi
```
Can it be changed to:
```
if [ "$OPTARG" == "deb" ]
then
        SUNSHINE_PACKAGE_EXTENSION=$OPTARG
elif [ "$OPTARG" == "rpm" ]
then
        SUNSHINE_PACKAGE_EXTENSION=$OPTARG
else
        echo "Package extension not supported: $OPTARG"
        echo "Falling back to default package extension: $SUNSHINE_PACKAGE_EXTENSION"
fi
```
It's a little bit longer/clunkier, but would this work?

**Comment by ReenigneArcher** [comment]
> @ReenigneArcher Got an idea.
> 
> Instead of using: (lines 55 - 62)
> 
> ```
> if `list_include_item "deb rpm" $OPTARG`
> then
> 	SUNSHINE_PACKAGE_EXTENSION=$OPTARG
> else
> 	echo "Package extension not supported: $OPTARG"
> 	echo "Falling back to default package extension: $SUNSHINE_PACKAGE_EXTENSION"
> fi
> ```
> 
> Can it be changed to:
> 
> ```
> if [ "$OPTARG" == "deb" ]
> then
>         SUNSHINE_PACKAGE_EXTENSION=$OPTARG
> elif [ "$OPTARG" == "rpm" ]
> then
>         SUNSHINE_PACKAGE_EXTENSION=$OPTARG
> else
>         echo "Package extension not supported: $OPTARG"
>         echo "Falling back to default package extension: $SUNSHINE_PACKAGE_EXTENSION"
> fi
> ```
> 
> It's a little bit longer/clunkier, but would this work?

This should work... Google failed me in this one.

**Comment by ReenigneArcher** [comment]
Same error... I don't understand where this is coming from... `logname: no login name`

**Comment by thatsysadmin** [comment]
I ran the script with the docker stuff commented out just now, and it seems like it isn't pulling in the `deb` or `rpm` arguments at the moment.

File: scripts/build-sunshine.sh
Comment (ReenigneArcher) [inline scripts/build-sunshine.sh:8]: Fix indent before `-e` ... sorry, my IDE was screwy here. Actually, don't know why the script is using tabs.

File: scripts/build-sunshine.sh
Comment (ReenigneArcher) [inline scripts/build-sunshine.sh:37]: This is the problem. 99% sure!

`while getopts ":dpuhc:e:s:n:" arg; do`

The first values (dpuhc) don't have values attached.

File: scripts/Dockerfile-debian
Comment (ReenigneArcher) [inline scripts/Dockerfile-debian:38]: While fixing stuff, add a blank line at the end

File: .github/workflows/CI.yml
Comment (ReenigneArcher) [inline .github/workflows/CI.yml:159]: While fixing stuff, just delete these 3 lines altogether.

File: gen-rpm.in
Comment (ReenigneArcher) [inline gen-rpm.in:162]: While fixing stuff, add a blank line at the end

File: scripts/Dockerfile-fedora_33
Comment (ReenigneArcher) [inline scripts/Dockerfile-fedora_33:27]: While fixing stuff, add a blank line at the end

File: scripts/Dockerfile-fedora_35
Comment (ReenigneArcher) [inline scripts/Dockerfile-fedora_35:32]: While fixing stuff, add a blank line at the end

File: scripts/Dockerfile-ubuntu_20_04
Comment (ReenigneArcher) [inline scripts/Dockerfile-ubuntu_20_04:44]: While fixing stuff, add a blank line at the end

File: scripts/Dockerfile-ubuntu_21_04
Comment (ReenigneArcher) [inline scripts/Dockerfile-ubuntu_21_04:37]: While fixing stuff, add a blank line at the end

File: scripts/Dockerfile-ubuntu_21_10
Comment (ReenigneArcher) [inline scripts/Dockerfile-ubuntu_21_10:37]: While fixing stuff, add a blank line at the end

**Review by ReenigneArcher — CHANGES_REQUESTED** [review]
Pretty sure I found the problem

File: scripts/build-sunshine.sh
Comment (thatsysadmin) [reply scripts/build-sunshine.sh:37]: I guess I didn't change that. Thought I did.

**Comment by thatsysadmin** [comment]
Do I need to pull all these changes into my branch?

File: scripts/build-sunshine.sh
Comment (thatsysadmin) [reply scripts/build-sunshine.sh:8]: Are you using VSCode? VSCode doesn't seem to tell you if it's spaces or tabs unless you select the text.

**Comment by ReenigneArcher** [comment]
This one I think I know...

```txt
docker: Error response from daemon: failed to create shim: OCI runtime create failed: container_linux.go:380: starting container process caused: exec: "/root/build.sh -rpm": stat /root/build.sh -rpm: no such file or directory: unknown.
time="2022-04-11T01:12:06Z" level=error msg="error waiting for container: context canceled"
```

One second and I'll update this comment.

**Comment by thatsysadmin** [comment]
Should be fixed now.

**Comment by thatsysadmin** [comment]
Aware of the script not being able to delete the rpmbuild folder. Taking a look.

**Comment by thatsysadmin** [comment]
@ReenigneArcher 

Ok, fixed. Try running the CI again.

File: scripts/build-sunshine.sh
Comment (ReenigneArcher) [inline scripts/build-sunshine.sh:119]: I missed this one...

```bash
docker cp $CONTAINER_NAME:/root/sunshine-build/package-deb/sunshine.$SUNSHINE_PACKAGE_EXTENSION "$BUILD_DIR/$CONTAINER_NAME.$SUNSHINE_PACKAGE_EXTENSION"
```

**Comment by thatsysadmin** [comment]
@ReenigneArcher 

Ok, fixed.

**Comment by ReenigneArcher** [comment]
Can you allow edits by maintainers?

File: scripts/build-sunshine.sh
Comment (ReenigneArcher) [inline scripts/build-sunshine.sh:114]: We need to make the rpm build here:

`/root/sunshine-build/package-rpm/sunshine.rpm`

Is that something simple to change? Even if just copying the rpm file to that location?

Comment (thatsysadmin) [reply scripts/build-sunshine.sh:114]: Can I add a `-d` to `gen-rpm` to retain compatability with running the script out of container?

Comment (ReenigneArcher) [reply scripts/build-sunshine.sh:119]: I think that's fine... Then in `build_private.sh` it will become `./gen-rpm -d` ?

File: scripts/build-private.sh
Comment (ReenigneArcher) [inline scripts/build-private.sh:48]: A couple of typos in here:

```bash
# Get preferred package format
if [ "$1" == "-rpm" ]
then
  echo "Packaging in .rpm format."
  ./gen-rpm
elif [ "$1" == "-deb" ]
then
  echo "Packaging in .deb format."
  ./gen-deb
else
  echo "Preferred packaging not specified."
  echo "Use -deb or -rpm to specify preferred package format."
  exit 1
fi
```

Comment (thatsysadmin) [reply scripts/build-private.sh:48]: Good catch.

File: scripts/build-sunshine.sh
Comment (thatsysadmin) [reply scripts/build-sunshine.sh:119]: Done.

**Comment by thatsysadmin** [comment]
@ReenigneArcher 

Can you try running it again?

File: gen-rpm.in
Comment (thatsysadmin) [reply gen-rpm.in]: Yeah, I forgot the whitespaces. Your changes look good though.

File: gen-rpm.in
Comment (ReenigneArcher) [inline gen-rpm.in:174]: ?

Comment (thatsysadmin) [reply gen-rpm.in:174]: My bad. Caught it after I pushed the commit.

File: gen-rpm.in
Comment (ReenigneArcher) [inline gen-rpm.in:12]: I'm guessing the variable isn't being set how we think.

Or do we need to wrap TRUE in quotes? `"TRUE"`

Just guessing. Not a bash expert.

Comment (ReenigneArcher) [reply gen-rpm.in:12]: Running your last commit now. That's the last one I can do tonight.

Appreciate your patience on this! We're close

Comment (thatsysadmin) [reply gen-rpm.in:12]: It shouldn't.

```
 h  /tmp  cat test.sh
if [ "$1" == "-d" ]; then
	export DOCKERSTATUS=TRUE
else
	export DOCKERSTATUS=FALSE
fi

echo $DOCKERSTATUS
 h  /tmp  bash test.sh
FALSE
 h  /tmp  bash test.sh -d
TRUE
 h  /tmp  
```

File: scripts/build-private.sh
Comment (ReenigneArcher) [inline scripts/build-private.sh:39]: Don't we need the `-d` argument? I thought that was already added?

Comment (thatsysadmin) [reply scripts/build-private.sh]: Yeah, we do.

**Comment by thatsysadmin** [comment]
Are you able to run the CI again? I added some simple debug outputs to see if the switch is working.

**Comment by ReenigneArcher** [comment]
Potentially just need to add a `mkdir -p` command

https://github.com/SunshineStream/Sunshine/runs/6000496185?check_suite_focus=true#step:4:272

https://linux.die.net/man/1/mkdir

**Comment by thatsysadmin** [comment]
Yeah, I thought so too.

File: gen-rpm.in
Comment (ReenigneArcher) [inline gen-rpm.in:168]: I think we want this. Right?

`mkdir -p ~/root/sunshine-build/package-rpm`

Comment (thatsysadmin) [reply gen-rpm.in:168]: No. `~` is `/root` because root is running it. `~/root/sunshine-build/package-rpm` assumes `/root/root/sunshine-build/package-rpm`.

Comment (ReenigneArcher) [reply gen-rpm.in]: My point was to add the `-p` argument, and don't include the file name, only the directory name.

Comment (thatsysadmin) [reply gen-rpm.in]: Yeah. Also didn't mean to put in the tilde. Whoops.

Comment (thatsysadmin) [reply gen-rpm.in]: > My point was to add the `-p` argument, and don't include the file name, only the directory name.

I read your comment too fast. Sorry.

Comment (ReenigneArcher) [reply gen-rpm.in]: And remove this `/sunshine.rpm` from the end

Comment (thatsysadmin) [reply gen-rpm.in]: Done. Had a long day today; you can tell I'm running on fumes.

Comment (ReenigneArcher) [reply gen-rpm.in]: > Done. Had a long day today; you can tell I'm running on fumes.

All good, same for me... Hoping to get this one over the goal line

**Comment by ReenigneArcher** [comment]
Success on fedora 35! Need to wait for the other jobs to finish before you can download the artifact

**Comment by thatsysadmin** [comment]
Thanks for your help. Couldn't have done it without you.

**Comment by ReenigneArcher** [comment]
> Thanks for your help. Couldn't have done it without you.

No problem. Workflows on github are a pain to test. Appreciate your patience with this, I think a lot of people would give up, haha.

P.S. I fixed the ubuntu 18.04 issue.

**Comment by ReenigneArcher** [comment]
Is this considered a success? The file within the archive is what will be published to releases as well.

image

**Comment by thatsysadmin** [comment]
That should be it, but let me test it first.

**Comment by thatsysadmin** [comment]
The package is good, and everything's in place. The only thing that isn't working is audio, which probably has something to do with pipewire/pulseaudio permissions because when the binary runs it gets an access denied, but the package itself works.

**Comment by ReenigneArcher** [comment]
> The package is good, and everything's in place. The only thing that isn't working on my machine is audio streaming, which probably has something to do with pipewire/pulseaudio permissions because when the binary runs it gets an access denied, but the package itself works.

Okay, is that something that needs to change here, or just a command the user has to run to grant the permission?

**Comment by thatsysadmin** [comment]
I'm not 100% sure, but if it was a command/a rulefile that I'm aware, I would have put it in already. It's especially intresting because I ran it as root.

**Review by ReenigneArcher — APPROVED** [review]

**Pull request merged**
